### ISSUE TITLE
Status of Bubble & Lantern

### ISSUE THREAD

**Issue by RogerLevy** [ForthHub/discussion#47]
Last September I posted about my ANS Forth game engine with brief mention of its ancillary namespacing system which I call Idioms.  Idioms are extensible gathered wordlists, with a convenient and complete declaration and switching grammar.  The engine is now called Lantern and it's sort of a hybrid of 3 engines which powered The Lady and 2 unreleased games, one of which was started over 10 years ago.  Thanks to idioms, the engine features an easy-to-use lexicon with a laser-like focus, without limiting the ability to provide more powerful features later on.  Or at least that's the goal.  I think it's going pretty well.

Instead of direct links, I'm just going to direct everyone to my profile - you can find them there.

Things were getting messy and uncertain last year, and I ended up taking a few months off from all programming due to burnout.  But I got a new part time job that let me resume work.  Many bugs were fixed (particularly with idioms) and the file structure was improved.  The in-game "IDE" got a mini-rehaul and I started developing a GUI system.  Having an in-game REPL is essential for fusing code and design, which is the direction I plan to go further in.

Bubble is pretty much stable.  For Lantern, I'm on the final lap to 1.0 Beta.  1.0 is not meant to be very big.  I have just a handful of features to complete and then I plan to shift gears to doing more game development within the defined boundaries (which still are pretty huge).  A lot of the functionality is already provided by Bubble - it's bringing it all together that takes some planning and decision making.  There will also be a handful of nice examples.  (I use those to develop and test functionality.)

This is the apex of my years of research and development in Forth!  The final destination.  No more abandoned experiments.  I'm excited to share it in its alpha phase and to enter a new chapter that should be marked by creativity and buttloads of fun.

**Comment by larsbrinkhoff** [comment]
Thank you, this certainly looks interesting.  I'm not much into game programming, or Windows, but this may change in the future.

I don't see much in the way of instructions in the Lantern repo.  Do those from Bubble apply?

**Comment by TG9541** [comment]
Roger, do you have docs on "idioms"?
By the way, abandoned experiments are a good thing: failure is a necessary result of research (the alternatives are "doing the obvious", or "pure luck").

**Comment by xieyuheng** [comment]
here are some docs about the namespacing system
:: https://github.com/RogerLevy/lantern/blob/master/bu/lib/idiom.f

**Comment by xieyuheng** [comment]
@RogerLevy 
since the namespace-system can be used as an OO-system,
it might be helpful to explain it
by comparing it with classical OO-system [maybe smalltalk ?],
concept by concept, advantage by advantage.

**Comment by RogerLevy** [comment]
I updated the dox at the top of idiom.f.

I can't compare it to anything except that it doesn't function like a conventional namespace system.  You are given a lot of freedom.  It's also not meant to make everything totally drop-in, like in other languages.

I do use it like a classical OO system sometimes.  It's essential for game object programming - without it you can't make larger games without getting very verbose.  But you are meant to compartmentalize by topic not necessarily data.  Only sometimes do those overlap.  If you treat them like classes you will get no benefit, it will probably make things harder.  Just use an OOF system then.

Advantages:  

- Helps avoid name collisions by allowing private words.  Don't overuse this feature though, it gets cumbersome if you need to test private words too much.  I use it mostly for extremely simple (1 or 2-letter) words, throwaways, or stuff that will never ever ever need to be tested or used outside the idiom's context.
- Helps avoid name collisions another way; you can "peel back" the search order to define a component that is unaware of the larger application at any point, more conveniently than with the stock ANS Forth facilities
- You package all the words used by a component together with it.  Eases testing.
- Is conducive to programming in "phases" that stack on top of each other
- Context is modal rather than bracketed; you don't need a closing bracket (or opening for that matter) and idioms can span multiple files.  You don't need to give each file its own idiom either.  Very permissive and flexible - therefore it takes a little practice and thought about how you should partition.

**Comment by RogerLevy** [comment]
@larsbrinkhoff Yep, install instructions for Lantern are the same as Bubble.  The documentation will take the form of examples at this stage - I'm prioritizing game development over documentation until sometime later this year, which will likely be around the point it's taken out of beta and made an official release.

**Comment by paraplegic** [comment]
Roger, your enthusiasm is infectious.  I'll be spending more time with this in a week or so, and thanks for the additional information on idioms ...

Rob Sciuk

**Comment by RogerLevy** [comment]
I've completed all the planned features for the 0.9 alpha milestone and decided to learn another game engine to make games in for the time being.  So the project is now on indefinite hiatus.  

Once the remaining problems regarding namespacing, object management, portability, and dependence on a commercial Forth are resolved I'm sure it could one day be a very strong engine.  I am done with working on these technical problems for now though.
